### PR TITLE
rp/uart: use lockfree ringbuffer.

### PR DESCRIPTION
**Pull request by Dirbaio**
This gets rid of another PeripheralMutex usage.

Extracted out of #980

### CODE REVIEW

File: embassy-hal-common/src/atomic_ring_buffer.rs [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        (unsafe { buf.add(start) }, n)
+    }
+
+    pub fn pop_done(&self, n: usize) {
+        trace!("  ringbuf: pop {:?}", n);
+
+        let start = self.0.start.load(Ordering::Relaxed);
+        self.0.start.store(self.0.wrap(start + n), Ordering::Release);
+    }
+}
```
Comment (huntc) [inline embassy-hal-common/src/atomic_ring_buffer.rs:156]: Would it be possible to get some tests for this? It looks quite testable. Thanks.

**Review by huntc — APPROVED** [review]
Thanks for this - just one comment, but LGTM.

**Review by lulf — APPROVED** [review]
LGTM, I really like the ergonomics of not having to manage the State type in the app code!

I agree with @huntc it would be nice to have some tests for the core ringbuffer part - it's the kind of code one can accidentally make bad changes too in the future. Also, if you have some references or explanation of why the barriers are valid, that would be helpful to any readers I think. Then again, it hopefully won't need to change...

**Comment by Dirbaio** [comment]
Thanks for the feedback! Added docs, tests, comments explaining the atomic orderings.

bors r+

**Pull request merged**